RecoCaloTools/Navigation/interface/CaloRectangle.h

```diff
@@ -0,0 +1,116 @@
+#ifndef RecoCaloTools_Navigation_CaloRectangle_H
+#define RecoCaloTools_Navigation_CaloRectangle_H
+
+#include "Geometry/CaloTopology/interface/CaloSubdetectorTopology.h"
+#include "Geometry/CaloTopology/interface/CaloTopology.h"
+
+/*
+ * CaloRectangle is a class to create a rectangular range of DetIds around a
+ * central DetId. Meant to be used for range-based loops to calculate cluster
+ * shape variables.
+ */
+
+struct CaloRectangle {
+    const int ixMin;
+    const int ixMax;
+    const int iyMin;
+    const int iyMax;
+
+    template<class T>
+    auto operator()(T home, CaloTopology const& topology);
+
+};
+
+inline auto makeCaloRectangle(int size) {
+    return CaloRectangle{-size, size, -size, size};
+}
+
+inline auto makeCaloRectangle(int sizeX, int sizeY) {
+    return CaloRectangle{-sizeX, sizeX, -sizeY, sizeY};
+}
+
+template<class T>
+T offsetBy(T start, CaloSubdetectorTopology const& topo, int dX, int dY)
+{
+    for(int x = 0; x < std::abs(dX) && start != T(0); x++) {
+        // east is eta in barrel
+        start = dX > 0 ? topo.goEast(start) : topo.goWest(start);
+    }
+
+    for(int y = 0; y < std::abs(dY) && start != T(0); y++) {
+        // north is phi in barrel
+        start = dY > 0 ? topo.goNorth(start) : topo.goSouth(start);
+    }
+    return start;
+}
+
+template<class T>
+class CaloRectangleRange {
+
+  public:
+
+    class Iterator {
+
+      public:
+
+        Iterator(T const& home, int ix, int iy, CaloRectangle const rectangle, CaloSubdetectorTopology const& topology)
+          : home_(home)
+          , rectangle_(rectangle)
+          , topology_(topology)
+          , ix_(ix)
+          , iy_(iy)
+        {}
+
+        Iterator& operator++() {
+            if(iy_ == rectangle_.iyMax) {
+                iy_ = rectangle_.iyMin;
+                ix_++;
+            } else ++iy_;
+            return *this;
+        }
+
+        int ix() const { return ix_; }
+        int iy() const { return iy_; }
+
```

Comment:
So a usability comment.  This is ieta, iphi in the barrel, not ix, iy which is just the endcap. 

This might cause user confusion. Obviously there is no good solution here as the variable name changes based on the subdector. In my private code, I've always used  iEtaOrIX()  / iPhiOrIY() to solve this issue. However I'm more than happy to hear other peoples opinions / suggestions on what is the best course of action here for the names.